larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,483 @@
+
+    //Set current list to be the same as before reclustering
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Pfo>(*this, initialPfoListName));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::BuildNewTwoDClusters(const Pfo *pPfoToRebuild, ClusterList &newClustersList)
+{
+    ClusterList clusterList2D;
+    LArPfoHelper::GetTwoDClusterList(pPfoToRebuild, clusterList2D);
+
+    for(const Cluster *const pTwoDCluster : clusterList2D)
+    {
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::RemoveFromPfo(*this, pPfoToRebuild, pTwoDCluster));
+        HitType hitType = LArClusterHelper::GetClusterHitType(pTwoDCluster);
+        std::string clusterListName(hitType == TPC_VIEW_U ? "ClustersU" : hitType == TPC_VIEW_V ? "ClustersV" : "ClustersW");
+
+        std::string initialListName="";
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Cluster>(*this, initialListName));
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Cluster>(*this, clusterListName));
+
+        std::string originalListName, fragmentListName;
+        ClusterList originalClusterList;
+        originalClusterList.push_back(pTwoDCluster);
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::InitializeFragmentation(*this, originalClusterList, originalListName, fragmentListName));
+    
+        const OrderedCaloHitList &twoDClusterOrderedCaloHitList(pTwoDCluster->GetOrderedCaloHitList());
+        OrderedCaloHitList leftoverCaloHitList = twoDClusterOrderedCaloHitList;
+    
+        int iCluster(0);
+        for(const Cluster *pNewCluster : newClustersList)
+        {
+            if (!pNewCluster)
+            {
+                std::cout << "Error: found null pointer in list of new clusters!" << std::endl;
+                continue;
+            }
+            PandoraContentApi::Cluster::Parameters parameters;
+            CaloHitList newClusterCaloHitList3D;
+            pNewCluster->GetOrderedCaloHitList().FillCaloHitList(newClusterCaloHitList3D); 
+    
+            for(const CaloHit *const p3DCaloHit : newClusterCaloHitList3D)
+            {
+                for (const OrderedCaloHitList::value_type &mapEntry : twoDClusterOrderedCaloHitList)
+                {
+                    for (const CaloHit *const pCaloHit : *mapEntry.second)
+                    {
+                        if(pCaloHit==static_cast<const CaloHit *>(p3DCaloHit->GetParentAddress())) 
+                        {
+                          parameters.m_caloHitList.push_back(static_cast<const CaloHit *>(pCaloHit));
+                          leftoverCaloHitList.Remove(pCaloHit);
+                        }
+                    }
+                }
```

Comment:
Might it make sense to construct a map between addresses and hits before hand so we don't have to loop over the full set of hits for each 3D hit, and just use the lookup instead? Failing that, you can presumably break from the inner most loop upon finding a match?

Reply:
Yeah I guess you're right. For now I've added the break as I am focusing on quick or major issues. Happy to revise this at a later stage.